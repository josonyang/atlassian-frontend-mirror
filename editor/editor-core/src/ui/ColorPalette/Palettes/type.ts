export type {
  PaletteColor,
  PaletteTooltipMessages,
} from '@atlaskit/editor-common/ui-color';
