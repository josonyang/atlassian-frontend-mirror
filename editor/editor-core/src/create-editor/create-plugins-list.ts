import type { CreateUIAnalyticsEvent } from '@atlaskit/analytics-next/types';
import type { EditorPluginInjectionAPI, EditorPresetBuilder } from '@atlaskit/editor-common/preset';
import type { ScrollGutterPluginOptions } from '@atlaskit/editor-plugins/base';
import type { BlockTypePluginOptions } from '@atlaskit/editor-plugins/block-type';

import type { DefaultPresetPluginOptions } from '../../src/presets/default';
import type { EditorPlugin, EditorProps } from '../types';
import type { EditorPluginFeatureProps } from '../types/editor-props';
import { isFullPage as fullPageCheck } from '../utils/is-full-page';

import { createFeatureFlagsFromProps } from './feature-flags-from-props';

const isCodeBlockAllowed = (options?: Pick<BlockTypePluginOptions, 'allowBlockType'>) => {
	const exclude =
		options && options.allowBlockType && options.allowBlockType.exclude
			? options.allowBlockType.exclude
			: [];

	return exclude.indexOf('codeBlock') === -1;
};

export function getScrollGutterOptions(props: EditorProps): ScrollGutterPluginOptions | undefined {
	const { appearance } = props;

	if (fullPageCheck(appearance)) {
		// Full Page appearance uses a scrollable div wrapper.
		return {
			getScrollElement: () => document.querySelector('.fabric-editor-popup-scroll-parent'),
		};
	}
	return undefined;
}

export function getDefaultPresetOptionsFromEditorProps(
	props: EditorProps,
	createAnalyticsEvent?: CreateUIAnalyticsEvent,
	// Omit placeholder since it's an existing prop in `DefaultPresetPluginOptions` and will get overidden there
): DefaultPresetPluginOptions & Omit<EditorPluginFeatureProps, 'placeholder'> {
	const appearance = props.appearance;

	const inputTracking = props.performanceTracking?.inputTracking;
	const cardOptions = props.linking?.smartLinks || props.smartLinks || props.UNSAFE_cards;

	return {
		...props,
		createAnalyticsEvent,
		typeAhead: {
			isMobile: false,
		},
		featureFlags: createFeatureFlagsFromProps(props),
		paste: {
			cardOptions,
			sanitizePrivateContent: props.sanitizePrivateContent,
		},
		base: {
			allowInlineCursorTarget: true,
			allowScrollGutter: getScrollGutterOptions(props),
			inputTracking,
			browserFreezeTracking: props.performanceTracking?.bFreezeTracking,
			ufo: createFeatureFlagsFromProps(props).ufo,
		},
		blockType: {
			lastNodeMustBeParagraph: appearance === 'comment' || appearance === 'chromeless',
			allowBlockType: props.allowBlockType,
			isUndoRedoButtonsEnabled: props.allowUndoRedoButtons,
		},
		placeholder: {
			placeholder: props.placeholder,
			placeholderBracketHint: props.placeholderBracketHint,
		},
		textFormatting: {
			...(props.textFormatting || {}),
			responsiveToolbarMenu:
				props.textFormatting?.responsiveToolbarMenu != null
					? props.textFormatting.responsiveToolbarMenu
					: props.allowUndoRedoButtons,
		},
		submitEditor: props.onSave,
		quickInsert: {
			enableElementBrowser: props.elementBrowser && props.elementBrowser.showModal,
			elementBrowserHelpUrl: props.elementBrowser && props.elementBrowser.helpUrl,
			disableDefaultItems: false,
			headless: false,
			emptyStateHandler: props.elementBrowser && props.elementBrowser.emptyStateHandler,
		},
		selection: { useLongPressSelection: false },
		hyperlinkOptions: {
			editorAppearance: props.appearance,
			linkPicker: props.linking?.linkPicker,
			onClickCallback: cardOptions?.onClickCallback,
			platform: 'web',
		},
		codeBlock: {
			...props.codeBlock,
			useLongPressSelection: false,
			allowCompositionInputOverride: false,
		},
	};
}

/**
 * Maps EditorProps to EditorPlugins
 *
 * Note: The order that presets are added determines
 * their placement in the editor toolbar.
 */
export default function createPluginsList(
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	preset: EditorPresetBuilder<any, any>,
	props: EditorProps,
	pluginInjectionAPI?: EditorPluginInjectionAPI,
): EditorPlugin[] {
	const excludes = new Set<string>();

	if (!isCodeBlockAllowed({ allowBlockType: props.allowBlockType })) {
		excludes.add('codeBlock');
	}
	return preset.build({ pluginInjectionAPI, excludePlugins: excludes });
}
