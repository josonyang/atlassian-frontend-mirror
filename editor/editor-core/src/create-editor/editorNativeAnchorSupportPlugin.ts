import type { NodeViewConstructor } from '@atlaskit/editor-common/lazy-node-view';
import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import type { Node as PMNode, Schema } from '@atlaskit/editor-prosemirror/model';
import { DOMSerializer } from '@atlaskit/editor-prosemirror/model';
import { PluginKey } from '@atlaskit/editor-prosemirror/state';
import { expValEquals } from '@atlaskit/tmp-editor-statsig/exp-val-equals';

const injectNodeViewNodeTypeList: readonly string[] = [
	'paragraph',
	'heading',
	'bulletList',
	'orderedList',
	'blockquote',
	'rule',
	'decisionList',
	'taskList',
	'layoutColumn',
	'hardBreak',
] as const;

const injectNodeViewNodeTypeConditionalList: readonly string[] = [
	...injectNodeViewNodeTypeList,
	// nodes that have custom nodeView behind experiment/FG
	'layoutSection',
] as const;

const key = new PluginKey('editorNativeAnchorSupportPlugin');

// Internal plugin to enable native anchor support in the editor.
// This plugin injects node views for specific node types (e.g., 'paragraph', 'heading')
// that do not already have a nodeView, allowing them to be rendered with native anchors when needed.
export const createEditorNativeAnchorSupportPlugin = (schema: Schema) => {
	const nodeViewEntries: [string, NodeViewConstructor][] = [];

	const injectList = expValEquals('advanced_layouts', 'isEnabled', true) && !expValEquals('platform_editor_breakout_resizing', 'isEnabled', true)
		? injectNodeViewNodeTypeList
		: injectNodeViewNodeTypeConditionalList;

	schema.spec.nodes.forEach((nodeName, nodeSpec) => {
		if (injectList.includes(nodeName)) {
			if (nodeSpec.toDOM && typeof nodeSpec.toDOM === 'function') {
				const toDOM = nodeSpec.toDOM.bind(nodeSpec);
				nodeViewEntries.push([
					nodeName,
					(node: PMNode) => DOMSerializer.renderSpec(document, toDOM(node)),
				]);
			}
		}
	});

	return new SafePlugin({
		key,
		props: {
			nodeViews: Object.fromEntries(nodeViewEntries),
		},
	});
};
