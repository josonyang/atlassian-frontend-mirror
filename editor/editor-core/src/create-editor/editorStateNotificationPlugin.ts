import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import {
	PluginKey,
	type ReadonlyTransaction,
	Transaction,
	type EditorState,
} from '@atlaskit/editor-prosemirror/state';
import { fg } from '@atlaskit/platform-feature-flags';

import { type EditorViewStateUpdatedCallbackProps } from '../types/editor-config';

const key = new PluginKey<EditorStateNotificationPluginState>('editorStateNotificationPlugin');

export type EditorStateNotificationPluginState = {
	latestTransaction: ReadonlyTransaction | undefined;
};

export const createEditorStateNotificationPlugin = (
	onEditorStateUpdated: (props: {
		oldEditorState: EditorState;
		newEditorState: EditorState;
	}) => void,
	onEditorViewStateUpdatedCallbacks: Array<{
		pluginName: string;
		callback: (props: EditorViewStateUpdatedCallbackProps) => void;
	}>,
) =>
	new SafePlugin<EditorStateNotificationPluginState>({
		key: key,
		state: {
			init() {
				return {
					latestTransaction: undefined,
				};
			},
			apply(tr) {
				return {
					latestTransaction: tr,
				};
			},
		},
		view: () => {
			return {
				update: (view, oldEditorState) => {
					const originalTransaction = key.getState(view.state)
						?.latestTransaction as unknown as Readonly<Transaction>;

					if (originalTransaction && fg('platform_editor_migrate_state_updates')) {
						onEditorViewStateUpdatedCallbacks.forEach((entry) => {
							entry.callback({
								originalTransaction,
								transactions: [originalTransaction as Transaction],
								oldEditorState,
								newEditorState: view.state,
							});
						});
					}
					onEditorStateUpdated({ oldEditorState, newEditorState: view.state });
				},
			};
		},
	});
