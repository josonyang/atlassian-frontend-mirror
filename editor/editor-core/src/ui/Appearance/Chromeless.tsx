/** @jsx jsx */
import React, { Fragment } from 'react';

import { css, jsx } from '@emotion/react';

import { useSharedPluginState } from '@atlaskit/editor-common/hooks';
import type { MaxContentSizePluginState } from '@atlaskit/editor-plugins/max-content-size';
import { token } from '@atlaskit/tokens';

import { usePresetContext } from '../../presets/context';
import type { EditorAppearance, EditorAppearanceComponentProps } from '../../types';
import { createEditorContentStyle } from '../ContentStyles';
import PluginSlot from '../PluginSlot';
import { scrollbarStyles } from '../styles';
import WithFlash from '../WithFlash';

const chromelessEditorStyles = css(
	{
		lineHeight: '20px',
		height: 'auto',
		overflowX: 'hidden',
		overflowY: 'auto',
	},
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
	scrollbarStyles,
	{
		maxWidth: 'inherit',
		boxSizing: 'border-box',
		wordWrap: 'break-word',
		// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
		'div > .ProseMirror': {
			outline: 'none',
			whiteSpace: 'pre-wrap',
			padding: 0,
			margin: 0,
			// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
			'& > :last-child': {
				paddingBottom: token('space.100', '0.5em'),
			},
		},
	},
);

const ContentArea = createEditorContentStyle();
ContentArea.displayName = 'ContentArea';

export default class Editor extends React.Component<EditorAppearanceComponentProps, any> {
	static displayName = 'ChromelessEditorAppearance';

	private appearance: EditorAppearance = 'chromeless';
	private containerElement: HTMLElement | null = null;

	private renderChrome = ({ maxContentSize }: { maxContentSize?: MaxContentSizePluginState }) => {
		const {
			editorDOMElement,
			editorView,
			editorActions,
			eventDispatcher,
			providerFactory,
			contentComponents,
			customContentComponents,
			maxHeight,
			minHeight = 30,
			popupsMountPoint,
			popupsBoundariesElement,
			popupsScrollableElement,
			disabled,
			dispatchAnalyticsEvent,
			pluginHooks,
			featureFlags,
		} = this.props;
		const maxContentSizeReached = Boolean(maxContentSize?.maxContentSizeReached);

		return (
			<WithFlash animate={maxContentSizeReached}>
				<div
					css={[
						chromelessEditorStyles,
						maxHeight &&
							// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
							css({
								// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
								maxHeight: `${maxHeight}px`,
							}),
						// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
						css({
							// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
							minHeight: `${minHeight}px`,
						}),
					]}
					data-testid="chromeless-editor"
					ref={(ref: HTMLElement | null) => (this.containerElement = ref)}
				>
					<ContentArea
						// eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
						className="ak-editor-content-area"
						featureFlags={featureFlags}
					>
						{customContentComponents && 'before' in customContentComponents
							? customContentComponents.before
							: customContentComponents}
						<PluginSlot
							editorView={editorView}
							editorActions={editorActions}
							eventDispatcher={eventDispatcher}
							providerFactory={providerFactory}
							appearance={this.appearance}
							items={contentComponents}
							popupsMountPoint={popupsMountPoint}
							popupsBoundariesElement={popupsBoundariesElement}
							popupsScrollableElement={popupsScrollableElement}
							containerElement={this.containerElement}
							disabled={!!disabled}
							dispatchAnalyticsEvent={dispatchAnalyticsEvent}
							wrapperElement={this.containerElement}
							pluginHooks={pluginHooks}
						/>
						{editorDOMElement}
						{customContentComponents && 'after' in customContentComponents
							? customContentComponents.after
							: null}
					</ContentArea>
				</div>
			</WithFlash>
		);
	};

	render() {
		return <RenderWithPluginState renderChrome={this.renderChrome} />;
	}
}

interface PluginStates {
	maxContentSize?: MaxContentSizePluginState;
}
interface RenderChromeProps {
	renderChrome: (props: PluginStates) => React.ReactNode;
}

function RenderWithPluginState({ renderChrome }: RenderChromeProps) {
	const api = usePresetContext();
	const { maxContentSizeState } = useSharedPluginState(api, ['maxContentSize']);

	return <Fragment>{renderChrome({ maxContentSize: maxContentSizeState })}</Fragment>;
}
