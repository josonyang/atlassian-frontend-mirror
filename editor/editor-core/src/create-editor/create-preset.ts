import type { NextEditorPlugin } from '@atlaskit/editor-common/types';

import createUniversalPreset from '../presets/universal';
import type { EditorPlugin, EditorProps } from '../types';

import { getDefaultPresetOptionsFromEditorProps } from './create-plugins-list';
import { createFeatureFlagsFromProps } from './feature-flags-from-props';

// Separate file, we should not accidentally import this into the `ComposableEditor`
// otherwise it will blow up the bundle size.
export function createPreset(props: EditorProps, prevProps?: EditorProps) {
	const preset = createUniversalPreset(
		props.appearance,
		getDefaultPresetOptionsFromEditorProps(props),
		createFeatureFlagsFromProps(props),
		prevProps?.appearance,
	);

	return withDangerouslyAppendPlugins(preset)(props.dangerouslyAppendPlugins?.__plugins);
}

function withDangerouslyAppendPlugins(
	preset: ReturnType<typeof createUniversalPreset>,
): (plugins: EditorPlugin[] | undefined) => ReturnType<typeof createUniversalPreset> {
	function createEditorNextPluginsFromDangerouslyAppended(
		plugins: EditorPlugin[],
	): NextEditorPlugin<any, any>[] {
		return plugins ? plugins.map((plugin) => () => plugin) : [];
	}

	return (editorPluginsToAppend?: EditorPlugin[]) => {
		if (!editorPluginsToAppend || editorPluginsToAppend.length === 0) {
			return preset;
		}

		const nextEditorPluginsToAppend =
			createEditorNextPluginsFromDangerouslyAppended(editorPluginsToAppend);

		const presetWithAppendedPlugins = nextEditorPluginsToAppend.reduce((acc, plugin) => {
			// These are dangerously appended, we don't need their type information leaking into
			// universal preset
			// @ts-expect-error
			return acc.add(plugin) as unknown as ReturnType<typeof createUniversalPreset>;
		}, preset);

		return presetWithAppendedPlugins;
	};
}
