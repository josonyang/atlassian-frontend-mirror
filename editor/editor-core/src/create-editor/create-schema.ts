import { createProseMirrorMetadata } from '@atlaskit/editor-common/prosemirror-dom-metadata';
import type {
	DOMOutputSpec,
	MarkSpec,
	NodeSpec,
	Mark as PMMark,
	Node as PMNode,
} from '@atlaskit/editor-prosemirror/model';
import { Schema } from '@atlaskit/editor-prosemirror/model';
import { fg } from '@atlaskit/platform-feature-flags';

import type { MarkConfig, NodeConfig } from '../types/pm-config';
import { sanitizeNodes } from '../utils/sanitizeNodes';

import { fixExcludes } from './create-editor';
import { sortByOrder } from './sort-by-order';

type toDOMType = (node: PMNode | PMMark) => DOMOutputSpec;

/**
 * ðŸ§± Internal Helper Function: Editor FE Platform
 *
 * Adds generic metadata attributes to a DOMOutputSpec array based on the provided node or mark.
 * This function ensures that the DOMOutputSpec is annotated with ProseMirror-specific metadata.
 *
 * @param {Object} params - Parameters object.
 * @param {PMNode | PMMark} params.nodeOrMark - The ProseMirror node or mark to extract metadata from.
 * @param {DOMOutputSpec} params.domSpec - The DOMOutputSpec to which attributes will be added.
 * @returns {DOMOutputSpec} The modified DOMOutputSpec with additional metadata.
 */
export const addMetadataAttributes = ({
	nodeOrMark,
	domSpec,
}: {
	nodeOrMark: PMNode | PMMark;
	domSpec: DOMOutputSpec;
}): DOMOutputSpec => {
	if (!Array.isArray(domSpec)) {
		return domSpec;
	}

	const maybeDefinedAttributes = domSpec[1];
	const metadata = createProseMirrorMetadata(nodeOrMark);
	const hasDefinedAttributes =
		typeof maybeDefinedAttributes === 'object' && !Array.isArray(maybeDefinedAttributes);

	if (hasDefinedAttributes) {
		domSpec[1] = Object.assign(maybeDefinedAttributes, metadata);
	} else {
		domSpec.splice(1, 0, metadata);
	}

	return domSpec;
};

/**
 * ðŸ§± Internal Helper Function: Editor FE Platform
 *
 * Wraps a `toDOM` function with a proxy that automatically adds metadata attributes
 * to the resulting DOMOutputSpec. This is useful for dynamically enhancing the output
 * of a `toDOM` function with additional context.
 *
 * @param {function(PMNode | PMMark): DOMOutputSpec} toDOM - The original `toDOM` function.
 * @returns {function(PMNode | PMMark): DOMOutputSpec} A proxied `toDOM` function that adds metadata attributes.
 */
export const wrapToDOMProxy = (toDOM: toDOMType): toDOMType => {
	return new Proxy(toDOM, {
		apply(target, thisArg, argumentsList) {
			const domSpec: ReturnType<toDOMType> = Reflect.apply(target, thisArg, argumentsList);

			if (!Array.isArray(domSpec)) {
				return domSpec;
			}

			const nodeOrMark = argumentsList[0];
			return addMetadataAttributes({ nodeOrMark, domSpec });
		},
	});
};

/**
 * ðŸ§± Internal Helper Function: Editor FE Platform
 *
 * Wraps a NodeSpec or MarkSpec object with a proxy to enhance its `toDOM` method.
 * This proxy automatically adds metadata attributes to the DOM output of the `toDOM` method,
 * enriching the DOM representation with additional ProseMirror-specific metadata.
 *
 * For nodes thats use NodeViews, you can find the implementation of those attributes on this file:
 * @see `packages/editor/editor-common/src/safe-plugin/index.ts`
 *
 * @template T
 * @param {T} spec - The NodeSpec or MarkSpec object to be wrapped.
 * @returns {T} A proxied NodeSpec or MarkSpec object where the `toDOM` method is enhanced
 * with metadata attributes.
 */
export const wrapNodeSpecProxy = <T extends NodeSpec | MarkSpec>(spec: T): T => {
	return new Proxy<T>(spec, {
		get(target, prop, receiver) {
			const result = Reflect.get(target, prop, receiver);

			if (prop === 'toDOM' && typeof result === 'function') {
				return wrapToDOMProxy(result);
			}

			return result;
		},
	});
};

export function createSchema(editorConfig: { marks: MarkConfig[]; nodes: NodeConfig[] }) {
	const marks = fixExcludes(
		editorConfig.marks.sort(sortByOrder('marks')).reduce(
			(acc, mark) => {
				if (fg('platform_editor_breakout_use_css')) {
					acc[mark.name] = wrapNodeSpecProxy(mark.mark);
				} else {
					acc[mark.name] = mark.mark;
				}
				return acc;
			},
			{} as { [nodeName: string]: MarkSpec },
		),
	);
	const nodes = sanitizeNodes(
		editorConfig.nodes.sort(sortByOrder('nodes')).reduce(
			(acc, node) => {
				if (fg('platform_editor_breakout_use_css')) {
					acc[node.name] = wrapNodeSpecProxy(node.node);
				} else {
					acc[node.name] = node.node;
				}
				return acc;
			},
			{} as { [nodeName: string]: NodeSpec },
		),
		marks,
	);

	return new Schema({ nodes, marks });
}
