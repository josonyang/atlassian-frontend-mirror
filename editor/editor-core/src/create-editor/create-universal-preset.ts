import type { EditorPlugin, NextEditorPlugin } from '@atlaskit/editor-common/types';

import createUniversalPresetInternal, {
	type InitialPluginConfiguration,
} from '../presets/universal';
import type { EditorProps } from '../types';

import { getDefaultPresetOptionsFromEditorProps } from './create-plugins-list';
import { createFeatureFlagsFromProps } from './feature-flags-from-props';

// Separate file, we should not accidentally import this into the `ComposableEditor`
// otherwise it will blow up the bundle size.
export function createUniversalPreset({
	props,
	prevProps,
	initialPluginConfiguration,
}: {
	props: EditorProps;
	prevProps?: EditorProps;
	initialPluginConfiguration?: InitialPluginConfiguration;
}) {
	const preset = createUniversalPresetInternal({
		appearance: props.appearance,
		props: getDefaultPresetOptionsFromEditorProps(props),
		initialPluginConfiguration: initialPluginConfiguration,
		featureFlags: createFeatureFlagsFromProps(props.featureFlags),
		prevAppearance: prevProps?.appearance,
	});

	return withDangerouslyAppendPlugins(preset)(props.dangerouslyAppendPlugins?.__plugins);
}

function withDangerouslyAppendPlugins(
	preset: ReturnType<typeof createUniversalPresetInternal>,
): (plugins: EditorPlugin[] | undefined) => ReturnType<typeof createUniversalPresetInternal> {
	function createEditorNextPluginsFromDangerouslyAppended(
		plugins: EditorPlugin[],
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
	): NextEditorPlugin<any, any>[] {
		return plugins ? plugins.map((plugin) => () => plugin) : [];
	}

	return (editorPluginsToAppend?: EditorPlugin[]) => {
		if (!editorPluginsToAppend || editorPluginsToAppend.length === 0) {
			return preset;
		}

		const nextEditorPluginsToAppend =
			createEditorNextPluginsFromDangerouslyAppended(editorPluginsToAppend);

		const presetWithAppendedPlugins = nextEditorPluginsToAppend.reduce((acc, plugin) => {
			// These are dangerously appended, we don't need their type information leaking into
			// universal preset
			// @ts-expect-error
			return acc.add(plugin) as unknown as ReturnType<typeof createUniversalPresetInternal>;
		}, preset);

		return presetWithAppendedPlugins;
	};
}
