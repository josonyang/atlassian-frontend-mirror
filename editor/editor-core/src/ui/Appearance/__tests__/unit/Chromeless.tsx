import React from 'react';

import { render } from '@testing-library/react';

import type { DocBuilder } from '@atlaskit/editor-common/types';
// eslint-disable-next-line import/no-extraneous-dependencies -- Removed import for fixing circular dependencies
import { createEditorFactory } from '@atlaskit/editor-test-helpers/create-editor';
// eslint-disable-next-line import/no-extraneous-dependencies -- Removed import for fixing circular dependencies
import { doc, p, panel } from '@atlaskit/editor-test-helpers/doc-builder';
import { eeTest } from '@atlaskit/tmp-editor-statsig/editor-experiments-test-utils';
import { ffTest } from '@atlassian/feature-flags-test-utils';

import { ChromelessEditorContainer } from '../../../Appearance/Chromeless';

describe('Chromeless editor', () => {
	afterEach(() => {
		jest.clearAllMocks();
	});

	const createEditor = createEditorFactory();
	const editor = (doc: DocBuilder) =>
		createEditor({
			doc,
			editorProps: {
				allowPanel: true,
				appearance: 'chromeless',
				quickInsert: true,
			},
		});

	it('should keep paragraph as the last node', async () => {
		const { editorView, typeAheadTool } = editor(doc(p('{<>}')));
		await typeAheadTool.searchQuickInsert('info')?.insert({ index: 0 });

		expect(editorView.state.doc).toEqualDocument(doc(panel({ panelType: 'info' })(p('')), p('')));
	});

	ffTest.on(
		'platform_fix_extra_space_last_line_comment_editor',
		'render correct overridden styles',
		() => {
			eeTest
				.describe('platform_editor_core_static_emotion', 'static emotion on')
				.variant(true, () => {
					it('should render correct overridden styles', () => {
						const { container } = render(
							<ChromelessEditorContainer minHeight={100} maxHeight={200}>
								<p>Hello world</p>
								<p>Hello world</p>
							</ChromelessEditorContainer>,
						);

						expect(container).toMatchSnapshot();
					});
				});
			eeTest
				.describe('platform_editor_core_static_emotion', 'static emotion off')
				.variant(false, () => {
					it('should render correct overridden styles even if static emotion is off', () => {
						const { container } = render(
							<ChromelessEditorContainer minHeight={100} maxHeight={200}>
								<p>Hello world</p>
								<p>Hello world</p>
							</ChromelessEditorContainer>,
						);

						expect(container).toMatchSnapshot();
					});
				});
		},
	);
});
