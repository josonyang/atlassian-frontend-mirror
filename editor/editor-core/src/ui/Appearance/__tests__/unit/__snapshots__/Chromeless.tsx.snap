// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chromeless editor platform_fix_extra_space_last_line_comment_editor is on render correct overridden styles should render correct overridden styles 1`] = `
.emotion-0 {
  line-height: 20px;
  height: auto;
  overflow-x: hidden;
  overflow-y: auto;
  max-width: inherit;
  box-sizing: border-box;
  word-wrap: break-word;
  -ms-overflow-style: -ms-autohiding-scrollbar;
}

.emotion-0 div>.ProseMirror {
  outline: none;
  white-space: pre-wrap;
  padding: 0;
  margin: 0;
}

.emotion-0 div>.ProseMirror>:last-child {
  padding-bottom: var(--ds-space-100, 0.5em);
}

.emotion-0::-webkit-scrollbar-corner {
  display: none;
}

.emotion-0::-webkit-scrollbar-thumb {
  background-color: var(--ds-background-neutral-subtle, #00000000);
}

.emotion-0:hover::-webkit-scrollbar-thumb {
  background-color: var(--ds-background-neutral-bold, #44546F);
  border-radius: 8px;
}

.emotion-0::-webkit-scrollbar-thumb:hover {
  background-color: var(--ds-background-neutral-bold-hovered, #2C3E5D);
}



.emotion-0 .ProseMirror>p:last-of-type {
  margin-bottom: var(--ds-space-0, 0px);
}

<div>
  <div
    class="emotion-0"
    data-testid="chromeless-editor"
    id="chromeless-editor"
    style="max-height: 200px; min-height: 100px;"
  >
    <p>
      Hello world
    </p>
    <p>
      Hello world
    </p>
  </div>
</div>
`;
